balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,253 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
 	}
 }
 
+// TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList verifies that pickfirst
+// correctly detects the end of the first happy eyeballs pass when the timer
+// causes pickfirst to reach the end of the address list and failures are
+// reported out of order.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	cc, rb, bm := setupPickFirstLeaf(t, 3,
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if got, want := holds[0].Wait(ctx), true; got != want {
+		t.Fatalf("hold[%q].Wait() = %t, want %t", addrs[0], got, want)
+	}
+
+	// Ensure no other servers are contacted.
+	if got, want := holds[1].IsStarted(), false; got != want {
+		t.Fatalf("holds[%q].IsStarted() = %t, want %t", addrs[1], got, want)
+	}
+
```

Comment:
Nit: remove this newline, since this is essentially part of the same logical code block under the comment `	// Ensure no other servers are contacted.`